Encoder: allow change of container from ebml to ogg with _pushEncoded…

fredericgermain opened the pull request:
I wanted to write a piece of code to change some webm, audio-only file I have to .opus

webm file are EBML file, same as matroska (https://www.matroska.org/technical/specs/index.html)

opus file are ogg file, with a small header OpusHead+OpusTags handled by the Encoder.prototype._writeHeader of node-opus

I only wanted a piece of code to write the small OpusHeah+Tags to my output ogg stream, without having reencoding. So I made extracted a new function, _pushEncodedBuffer, to do this.

Note that it should probably be named as a public funtion as it is used externally in my use-case...

Here is the piece of code I'm using that do the job. I used fixed parameter for vorbis, there is probably a way to find them in the EBML data....

        const ebml = require('ebml')
        const ebmlBlock = require('ebml-block')
        const opus = require('node-opus');
        const ogg = require( 'ogg' );

        const oggEncoder = new ogg.Encoder();
        const opusEncoder = new opus.Encoder(48000, 2);

        const oggStream = oggEncoder.stream();
        const oggPipe = opusEncoder.pipe( oggStream );
        const decoder = new ebml.Decoder();

        webmStream
          .pipe(decoder)
          .pipe(new stream.Transform({
            transform: function(chunk, encoding, callback) {
              if (chunk[0] === 'tag' && (chunk[1].name === 'Block' || chunk[1].name === 'SimpleBlock')) {
                var webmTransform = this;
                var block = ebmlBlock(chunk[1].data)
                var encoded = block.frames[0];
                opusEncoder._pushEncodedBuffer(encoded);
              }
              callback();
            },
            writableObjectMode: true,
          }))

          var p = oggEncoder
          .pipe(fs.createWriteStream('audio.opus'))

Rantanen commented:
Had some other stuff to take care during the Easter holidays. I should get around to looking at this during this week.

Usually I've merged things in with a quick review, but this one needs a bit more in depth one as I can't tell by a glance how it affects the current code paths - ie. whether it has a chance of breaking existing code depending on the public methods.

Furthermore node-opus is primarily about the libopus bindings and this one completely skips those, so it's not on the highest priority for node-opus.

That said, I'm very interested in solving your use case as well and hopefully I'll feel confident enough on this one so that it won't break the existing uses. Even if it seems like there might be a chance for that, I'll see if I can come up with a way to maintain those.

So what I'm trying to say is sorry for the delays, but it's going to take a day or two more (hopefully not more than that!)

fredericgermain commented:
Completely understandable :)

I checked where this header  OpusHead+OpusTags is handled officially in C code. It's in the opus-tools package it seems (src/opus_header.h), and this package do not export any library to wrap.

So I think it would still make make sense to keep it here in node-opus. We could create a node-opus-tools but that seems really over complicated...

fredericgermain commented:
Hi again,

So I worked on node-opus encoder these last days to make it fit for my project.
I pushed it [there](https://github.com/fredericgermain/node-opus/tree/split-streams), but it's open to talk of course.

Encoder.js: keep former API, split the logic between OpusEncoderStream.js (PCM -> Encoded buffer) and OpusFileWriterStream (Encoded buffer -> ogg packet stream)

StreamConverter : A utility class which take a stream in input, and output a opus file stream. It can do : 
- a webm/mkv to opus convertion without reencoding.
- otherwise, it use ffmpeg program to decode to PCM then reencode & repack

VorbisComments: add meta tag writer on the output opus file. I couldn't find any on npm.js... maybe to merge to npmjs vorbis-parser (I really liked that parser)?

I fixed the end of file encoding (hopefully) : the last pcm data were not encoded
- in case of reencode, last pcm data is paded with 0 and encoded (crash encoder if not padded)
- in case of webm->opusfile, we pass the DiscardPadding information
- we take into account that information in the last ogg packet of the stream

That's quite a lot of things to test.. I tested on a 80+ files library I used (nearly 70% webm), I don't have any important problem when I listen to them.

Rantanen commented:
Hey!

Sorry I managed to forget this again. Good news though! I finally felt guilty for this pending and ended up adding some unit tests to the project.

While these are self tests (ie, they use the implementation itself for validation) they should catch any changes in the outputs and there were none here so I felt confident merging it in. The code also seemed like it shouldn't cause any.

I'll have a look at your branch as well while I'm at it.

Rantanen commented:
I can definitely get behind the idea of splitting `Encoder` -> `OpusEncoderStream`, which is a stream for converting PCM into Opus frames and `OggWriteStream`, which is a stream for converting the audio frames into an ogg file stream. (I'd mention the write-bit in the ogg stream name as in the future we'd also like to mirror this in the Decoder side with `OpusDecoderStream` and `OggReadStream`).

Couple of comments on the branch.

`Encoder._transform` causes massive back pressure on the encoder stream in case the output is slower than input (such as encoding files from local drive and uploading them over the network). The signaling of `done()` should take into account the encoder stream state. Something such as below - which I typed up based on a quick glance at the documentation - didn't actually test whether it's in any way correct.

```js
var writeMore = this.encoderStream.write( pcmbuf );

// Signal done since we can write more immediately.
if( writeMore )
    return done();

// Stream is full. Signal done only after the stream has drained.
this.encoderStream.once( 'drain', done );
```

`new Encoder()` returns non-`Encoder` instances if `type !== 'pcm'`. This yields situation where `new Encoder() instanceof Encoder` is false, which is a bit confusing.

the Encoder constructor should act as a constructor. This means the encoding always ends up going through the `Encoder::_transform()`. I would prefer if the constructor set the encoder up in the following manner (skipping error checks, etc. for brevity):

```js
// Set up the ogg stream as the output stream always.
var output = new OpusFileOggStream( ... );
output.on( 'data', function( chunk ) { this.push( chunk ); }.bind( this ) );

// Input stream depends on the type.
if( type === 'pcm' ) {

    // pcm uses opus encoder stream piped to output.
    this.input = new OpusEncoderStream( ... );
    this.input.pipe( output );

} else if( type === 'encoded' ) {

    // encoded uses the ogg stream as the input as well.
    this.input = output;
}
```

Concerning the `flush_done` state. Given the `WritableStream::end()` takes a callback parameter - can't the `_flush` be implemented with...

```js
Encoder.prototype._flush = function( done ) { this.encoderStream.end( done ); }
```

The `StreamConverter.js` has various issues and I believe it might be better to separate this to a different module:

- Various dependencies that are not present in package.json: `ogg`, `debug`, `ebml`, `ebml-block`. Also I don't think `node-opus` - which is a low level library for handling the opus codec should depend on `ebml`, etc. just for one specific use case. It's worth remembering we're not depending on `ogg` either even if we are `ogg` compatible.
- `const` and `=>` are modern Node features. `node-opus` is compatible with ancient Node.js versions down to `0.10.x` which doesn't support these.
